src/main/java/us/abstracta/wiresham/VirtualTcpService.java

```
@@ -124,24 +155,72 @@ private synchronized void removeClient(ConnectionFlowDriver connectionDriver) {
-    clientExecutorService.awaitTermination(timeoutMillis, TimeUnit.MILLISECONDS);
-    serverExecutorService.shutdownNow();
-    if (!serverExecutorService.awaitTermination(timeoutMillis, TimeUnit.MILLISECONDS)) {
-      LOG.warn("Server thread didn't stop after {} millis", timeoutMillis);
+    if (!clientExecutorService.awaitTermination(timeoutMillis, TimeUnit.MILLISECONDS)) {
+      clientExecutorService.shutdownNow();
     }
   }
 
+  private FlowConnectionProvider buildFlowConnectionProvider() {
+    return new FlowConnectionProvider() {
+      public final Map<Integer, CompletableFuture<FlowConnection>> map = new ConcurrentHashMap<>();
+
+      @Override
+      public FlowConnection get(int port) throws ExecutionException, InterruptedException {
+        return map.get(port).get();
+      }
+
+      @Override
+      public void init(List<Integer> ports, FlowConnection flowConnection) {
+        map.clear();
+        for (Integer port : ports) {
+          map.put(port, new CompletableFuture<>());
+        }
+        CompletableFuture<FlowConnection> completedFuture = new CompletableFuture<>();
+        completedFuture.complete(flowConnection);
+        map.put(flowConnection.getPort(), completedFuture);
+      }
+
+      @Override
+      public void assignFlowConnection(int port, FlowConnection flowConnection) {
+        map.get(port).complete(flowConnection);
+      }
+
+      @Override
+      public boolean requiresFlowConnection(int port) {
+        return !map.get(port).isDone();
+      }
+
+      @Override
+      public void closeConnections() throws IOException {
+        for (CompletableFuture<FlowConnection> value : map.values()) {
+          try {
+            value.get(CLOSE_SOCKETS_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS).close();
+          } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            //Do nothing since there might be a socket that was not used therefore, connection was 
```

Are you trying to overcome the issue of not having cancelled the completable? Sounds like you may be ignoring exceptions that you shouldn't (too many reasons for such exceptions)?